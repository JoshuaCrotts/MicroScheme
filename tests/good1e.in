

// A recursive powering function, using several more operations. One
// important point here is that recursive functions must know about the
// function being defined before the definition is complete (in other words
// if you wait until exitFnDecl or something similar to actually declare
// the function signature, then it won't recognize it inside the function
// body for recursion!

int pow(int a, int b) {
    int halfb;
    int recurse;
    if (b == 0)
        return 1;

    recurse = pow(a, b/2);
    if (b%2 == 1)
        return recurse*recurse*a;
    else
        return recurse*recurse;
}
        

void main() {
    printd(pow(2, 6));
}

    
