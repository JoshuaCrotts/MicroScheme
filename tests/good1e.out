void SEQ (
  void FNDEF pow int (int,int) (
    void SEQ (
      void PDECL a (int),
      void PDECL b (int),
      void DECL halfb (int),
      void DECL recurse (int),
      void IF (
        int BINOP('==') (
          int ID b,
          int LIT = 0),
        void RETURN (
          int LIT = 1)),
      int ASN (
        int ID recurse,
        int FNCALL pow (
          int ID a,
          int BINOP('/') (
            int ID b,
            int LIT = 2))),
      void IF (
        int BINOP('==') (
          int BINOP('%') (
            int ID b,
            int LIT = 2),
          int LIT = 1),
        void RETURN (
          int BINOP('*') (
            int BINOP('*') (
              int ID recurse,
              int ID recurse),
            int ID a)),
        void RETURN (
          int BINOP('*') (
            int ID recurse,
            int ID recurse))))),
  void FNDEF main void () (
    void SEQ (
      void FNCALL printd (
        int FNCALL pow (
          int LIT = 2,
          int LIT = 6)))))
